File: src/main/java/no/nav/bidrag/commons/web/test/HttpHeaderTestRestTemplate.java
package no.nav.bidrag.commons.web.test;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.Stack;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;

public class HttpHeaderTestRestTemplate {

  private final Stack<String[]> headersForSingleCallbacks = new Stack<>();
  private final Map<String, ValueGenerator> valueGenerators = new HashMap<>();
  private final TestRestTemplate testRestTemplate;

  public HttpHeaderTestRestTemplate(TestRestTemplate testRestTemplate) {
    this.testRestTemplate = testRestTemplate;
  }

  public <T> ResponseEntity<T> exchange(String url, HttpMethod httpMethod, HttpEntity<?> httpEntity, Class<T> responseClass) {
    return testRestTemplate.exchange(url, httpMethod, newEntityWithaddedHeaders(httpEntity), responseClass);
  }

  public <T> ResponseEntity<T> exchange(String url, HttpMethod httpMethod, HttpEntity<?> httpEntity, ParameterizedTypeReference<T> typeReference) {
    return testRestTemplate.exchange(url, httpMethod, newEntityWithaddedHeaders(httpEntity), typeReference);
  }

  private HttpEntity<?> newEntityWithaddedHeaders(HttpEntity<?> httpEntity) {
    HttpHeaders tempHeaders = new HttpHeaders();
    valueGenerators.forEach((key, value) -> tempHeaders.add(key, value.generate()));

    while (!headersForSingleCallbacks.empty()) {
      String[] headerWithValue = headersForSingleCallbacks.pop();
      tempHeaders.add(headerWithValue[0], headerWithValue[1]);
    }

    Optional.ofNullable(httpEntity).ifPresent(entity -> tempHeaders.putAll(entity.getHeaders()));

    return new HttpEntity<>(
        Optional.ofNullable(httpEntity).map(HttpEntity::getBody).orElse(null),
        tempHeaders
    );
  }

  public void add(String headerName, ValueGenerator valueGenerator) {
    valueGenerators.put(headerName, valueGenerator);
  }

  public void addHeaderForSingleHttpEntityCallback(String headerName, String headerValue) {
    String[] headerWithValue = {headerName, headerValue};
    headersForSingleCallbacks.push(headerWithValue);
  }

  @FunctionalInterface
  public interface ValueGenerator {

    String generate();
  }
}

File: src/main/kotlin/no/nav/bidrag/commons/web/test/HttpHeaderTestRestTemplate.kt
package no.nav.bidrag.commons.web.test

import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.test.web.client.exchange
import org.springframework.core.ParameterizedTypeReference
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.ResponseEntity
import java.net.URI
import java.util.*


class HttpHeaderTestRestTemplate(val testRestTemplate: TestRestTemplate) {

    private val headersForSingleCallbacks = Stack<Pair<String, String>>()
    private val valueGenerators: MutableMap<String, ValueGenerator> = HashMap()
    fun <T> exchange(url: String?, httpMethod: HttpMethod?, httpEntity: HttpEntity<*>, responseClass: Class<T>?): ResponseEntity<T> {
        return testRestTemplate.exchange(url, httpMethod, newEntityWithAddedHeaders(httpEntity), responseClass)
    }

    fun <T> exchange(
        url: String?,
        httpMethod: HttpMethod?,
        httpEntity: HttpEntity<*>?,
        typeReference: ParameterizedTypeReference<T>?
    ): ResponseEntity<T> {
        return testRestTemplate.exchange(url, httpMethod, newEntityWithAddedHeaders(httpEntity), typeReference)
    }

    fun <T> postForEntity(url: String?, httpEntity: HttpEntity<*>, responseClass: Class<T>?): ResponseEntity<T> {
        return testRestTemplate.postForEntity(url, newEntityWithAddedHeaders(httpEntity), responseClass)
    }

    inline fun <reified T : Any> getForEntity(uri: URI, request: Any? = null): ResponseEntity<T> {
        return testRestTemplate.exchange(uri, HttpMethod.GET, newEntityWithAddedHeaders(request))
    }

    inline fun <reified T : Any> putForEntity(uri: URI, request: Any? = null): ResponseEntity<T> {
        return testRestTemplate.exchange(uri, HttpMethod.PUT, newEntityWithAddedHeaders(request))
    }

    inline fun <reified T : Any> postForEntity(uri: URI, request: Any? = null): ResponseEntity<T> {
        return testRestTemplate.exchange(uri, HttpMethod.POST, newEntityWithAddedHeaders(request))
    }

    fun newEntityWithAddedHeaders(request: Any?): HttpEntity<*> {
        val tempHeaders = HttpHeaders().apply {
            valueGenerators.forEach { add(it.key, it.value.generate()) }
        }

        while (!headersForSingleCallbacks.empty()) {
            val headerWithValue = headersForSingleCallbacks.pop()
            tempHeaders.add(headerWithValue.first, headerWithValue.second)
        }

        if (request is HttpEntity<*>) {
            return HttpEntity(request.body, tempHeaders.apply { putAll(request.headers) })
        }
        return HttpEntity(request, tempHeaders)
    }

    fun add(headerName: String, valueGenerator: ValueGenerator) {
        valueGenerators[headerName] = valueGenerator
    }

    fun addHeaderForSingleHttpEntityCallback(headerName: String, headerValue: String) {
        val headerWithValue = headerName to headerValue
        headersForSingleCallbacks.push(headerWithValue)
    }

    fun interface ValueGenerator {
        fun generate(): String?
    }
}